src/ui.rs

```diff
@@ -259,14 +259,21 @@ impl Ui {
             clone!(shell_ref => move |args| set_background(&*shell_ref, args)),
         );
 
+        shell.state.borrow().subscribe(
+            SubscriptionKey::from("VimLeave"),
+            &["v:exiting"],
+            clone!(shell_ref => move |args| set_exit_status(&*shell_ref, args)),
+        );
+
         window.connect_delete_event(clone!(comps_ref, shell_ref => move |_, _| {
             gtk_delete(&*comps_ref, &*shell_ref)
         }));
 
         shell.grab_focus();
 
-        shell.set_detach_cb(Some(clone!(comps_ref => move || {
+        shell.set_detach_cb(Some(clone!(comps_ref => move |status| {
             glib::idle_add_once(clone!(comps_ref => move || comps_ref.borrow().close_window()));
+            std::process::exit(status);
```

Comment:
Not sure this is the right way to go about this. Unlike the neovim-gtk of old, [we actually attempt to fully execute both `VimPreLeave` and `VimLeave` when exiting neovim-gtk](https://github.com/Lyude/neovim-gtk/blob/main/src/shell.rs#L1169) (which calls down to [`shutdown()`](https://github.com/Lyude/neovim-gtk/blob/main/src/nvim/mod.rs#L279) which I'm fairly sure an early exit here via `std::process::exit(…)` would skip. The main reason for doing this is that many plugins will have cleanup routines hooked up to these events, and exiting without executing them could result in things such as leaking plugin processes (YouCompleteMe was a good example of this).
Doing a bit of research, it seems the "proper" way of doing this would likely be to do it through the `GApplication` abstraction that GIO provides, since we use that in `src/main.rs`. [See here](https://gtk-rs.org/gtk-rs-core/stable/0.14/docs/gio/struct.ApplicationCommandLine.html), we would likely want to hook up to the command-line event, and use it to stash a copy of the `ApplicationCommandLine` into `Shell` that later from somewhere like `detach_ui()` we can pass the correct error code.

Reply:
I've gone ahead and used the `ApplicationCommandLine` class.  I noted that it seems to work differently than the documentation led me to believe, and perhaps it also works differently than you thought it worked, but I've documented that in the new commit message.

I've also gone ahead and made some relatively extensive changes to `main.rs` because those are required for us to use the new callback.  Again, most of this is documented in the commit message.

Our exit call is now at the end of the program, which I think is really the ideal place for it.